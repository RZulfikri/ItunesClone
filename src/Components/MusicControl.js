import React, {PureComponent} from 'react';
import {Animated, StyleSheet, TouchableOpacity, View} from 'react-native';
import Colors from '../Themes/Colors';
import Icons from 'react-native-vector-icons/MaterialCommunityIcons';
import TrackPlayer, {STATE_PAUSED} from 'react-native-track-player';
import ProgressBar from './ProgressBar';
import ActiveSongItem from './ActiveSongItem';
import ItunesActions from '../Redux/ItunesRedux';
import {connect} from 'react-redux';

const MAX_HEIGHT = 100;
const MIN_HEIGHT = 0;

const styles = StyleSheet.create({
  container: {
    borderTopWidth: 1,
    borderTopColor: Colors.border,
    paddingHorizontal: 15,
    overflow: 'hidden',
    justifyContent: 'center',
  },
  topContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  actionContainer: {
    width: 50,
    height: 40,
  },
  playActionContainer: {marginHorizontal: 10},
  bottomContainer: {},
  infoContainer: {
    flex: 1,
  },
  playerContainer: {
    flex: 1,
    flexDirection: 'row',
  },
});

/**
 * This component is used as music player/control.
 * the feature is play, pause, stop, seek, backward and forward music/song
 */

class MusicControl extends PureComponent {
  activeSong;

  constructor(props) {
    super(props);
    this.state = {
      isPause: false,
      isStop: false,
      containerHeight: new Animated.Value(0),
    };

    this.show = this.show.bind(this);
    this.hide = this.hide.bind(this);
    this.play = this.play.bind(this);
    this.pause = this.pause.bind(this);
    this.stop = this.stop.bind(this);
    this.backward = this.backward.bind(this);
    this.forward = this.forward.bind(this);
    this.seek = this.seek.bind(this);
  }

  async componentDidMount() {
    const {setRef} = this.props;

    // action to passing function from child to parent component
    if (typeof setRef === 'function') {
      setRef({
        show: this.show,
        hide: this.hide,
        play: this.play,
        pause: this.pause,
        stop: this.stop,
      });
    }

    // an event listener for player playback state
    this.playbackListener = await TrackPlayer.addEventListener(
      'playback-state',
      async ({state}) => {
        if (state === STATE_PAUSED) {
          const position = await TrackPlayer.getPosition();
          const duration = await TrackPlayer.getDuration();
          if (duration > 0 && position > 0) {
            if (position === duration) {
              this.setState({isPause: true, isStop: true});
            } else {
              this.setState({isPause: true, isStop: false});
            }
          }
        }
      },
    );
  }

  async componentWillUnmount() {
    await TrackPlayer.stop();
    await TrackPlayer.destroy();
    this.playbackListener.remove();
  }

  // action to start show animation
  show() {
    Animated.timing(this.state.containerHeight, {
      toValue: MAX_HEIGHT,
      duration: 300,
      useNativeDriver: false,
    }).start(({finished}) => {
      /* completion callback */
      if (finished) {
        this.isShown = true;
      }
    });
  }

  // action to start hide animation
  hide() {
    Animated.timing(this.state.containerHeight, {
      toValue: MIN_HEIGHT,
      duration: 300,
      useNativeDriver: false,
    }).start(({finished}) => {
      /* completion callback */
      if (finished) {
        this.isShow = false;
      }
    });
  }

  // function to play/resume music
  play(data, force) {
    this.activeSong = data;
    if (this.state.isStop || force) {
      this.setState({isPause: false, isStop: false}, async () => {
        await TrackPlayer.stop();
        await TrackPlayer.setupPlayer();

        // Add a track to the queue
        await TrackPlayer.add({
          id: this.activeSong.get('trackId'),
          url: this.activeSong.get('previewUrl'),
          title: this.activeSong.get('trackName'),
          artist: this.activeSong.get('artistName'),
          artwork: this.activeSong.get('artworkUrl60'),
        });

        // Start playing it
        await TrackPlayer.play();
      });
    } else {
      this.setState({isPause: false, isStop: false}, async () => {
        await TrackPlayer.play();
      });
    }
  }

  // function to pause music player
  pause() {
    this.setState({isPause: true}, async () => {
      await TrackPlayer.pause();
    });
  }

  // function to stop music player
  stop() {
    this.setState({isPause: false, isStop: true}, async () => {
      this.activeSong = undefined;
      await TrackPlayer.stop();
    });
  }

  // function to seek playe based on slider position
  async seek(position) {
    this.setState({isPause: false}, async () => {
      const duration = await TrackPlayer.getDuration();
      await TrackPlayer.seekTo(position * duration);
      await TrackPlayer.play();
    });
  }

  // function to backward music
  backward() {
    const {setBackward} = this.props;
    setBackward();
  }

  // function to forward music
  forward() {
    const {setForward} = this.props;
    setForward();
  }

  render() {
    const {isPause, containerHeight} = this.state;
    return (
      <Animated.View style={[styles.container, {height: containerHeight}]}>
        <View style={styles.topContainer}>
          <View style={styles.infoContainer}>
            <ActiveSongItem />
          </View>
          <View style={styles.playerContainer}>
            <TouchableOpacity
              style={styles.actionContainer}
              activeOpacity={0.8}
              onPress={this.backward}>
              <Icons name={'skip-backward'} size={35} />
            </TouchableOpacity>
            <TouchableOpacity
              style={[styles.actionContainer, styles.playActionContainer]}
              activeOpacity={0.8}
              onPress={isPause ? () => this.play(this.activeSong) : this.pause}>
              <Icons name={isPause ? 'play' : 'pause'} size={35} />
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.actionContainer}
              activeOpacity={0.8}
              onPress={this.forward}>
              <Icons name={'skip-forward'} size={35} />
            </TouchableOpacity>
          </View>
        </View>
        <View style={styles.bottomContainer}>
          <ProgressBar onPause={this.pause} onSeek={this.seek} />
        </View>
      </Animated.View>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    setBackward: () => dispatch(ItunesActions.setBackward()),
    setForward: () => dispatch(ItunesActions.setForward()),
  };
};

export default connect(null, mapDispatchToProps)(MusicControl);
