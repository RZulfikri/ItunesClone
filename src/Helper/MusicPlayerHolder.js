class MusicPlayerHolder {
  player = undefined;
  activeSong = undefined;
  isShow = false;

  constructor() {
    this.setInstance = this.setInstance.bind(this);
  }

  setInstance(instance) {
    this.player = instance;
  }

  hide() {
    this.player.hide();
    this.isShow = false;
  }

  play(data) {
    if (!this.isShow) {
      this.player.show();
      this.isShow = true;
    }
    this.activeSong = data;
    this.player.play(data, true);
  }

  pause() {
    this.player.pause();
  }

  stop() {
    this.player.hide();
    this.isShow = false;
    this.activeSong = undefined;
    this.player.stop();
  }
}

export default new MusicPlayerHolder();
