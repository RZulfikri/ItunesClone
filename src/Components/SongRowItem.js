import React, {PureComponent} from 'react';
import {Image, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import MusicPlayerHolder from '../Helper/MusicPlayerHolder';
import Colors from '../Themes/Colors';
import ItunesActions from '../Redux/ItunesRedux';
import {connect} from 'react-redux';
import ActiveSongIndicator from './ActiveSongIndicator';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    padding: 15,
    borderBottomWidth: 1,
    borderBottomColor: Colors.border,
    alignItems: 'center',
  },
  infoContainer: {
    flex: 1,
    marginLeft: 10,
  },
  coverImage: {
    width: 50,
    height: 50,
  },
  songName: {
    fontSize: 14,
    fontWeight: 'bold',
  },
  songArtist: {
    fontSize: 12,
  },
  songAlbum: {
    fontSize: 10,
  },
});

class SongRowItem extends PureComponent {
  constructor(props) {
    super(props);
    this.onPressItem = this.onPressItem.bind(this);
  }

  onPressItem() {
    const {data, setActiveSong} = this.props;
    MusicPlayerHolder.play(data);
    setActiveSong(data.toJS());
  }

  render() {
    const {data} = this.props;
    return (
      <TouchableOpacity
        style={styles.container}
        activeOpacity={0.8}
        onPress={this.onPressItem}>
        <Image
          source={{uri: data.get('artworkUrl100')}}
          resizeMethod={'resize'}
          resizeMode={'cover'}
          style={styles.coverImage}
        />
        <View style={styles.infoContainer}>
          <Text style={styles.songName}>{data.get('trackName')}</Text>
          <Text style={styles.songArtist}>{data.get('artistName')}</Text>
          <Text style={styles.songAlbum}>{data.get('collectionName')}</Text>
        </View>
        <View>
          <ActiveSongIndicator data={data} />
        </View>
      </TouchableOpacity>
    );
  }
}

const mapDisPatchToProps = dispatch => {
  return {
    setActiveSong: params => dispatch(ItunesActions.setActiveSong(params)),
  };
};

export default connect(null, mapDisPatchToProps)(SongRowItem);
