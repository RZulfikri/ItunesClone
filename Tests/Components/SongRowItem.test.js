import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {fromJS, Map} from 'immutable';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import SongRowItem from '../../src/Components/SongRowItem';
import MusicPlayerHolder from '../../src/Helper/MusicPlayerHolder';
import MusicControl, {mockPlay} from '../__mocks__/MusicControl';
import ItunesActions from '../../src/Redux/ItunesRedux';

const mockStore = configureStore([]);

describe('SongRowItem React-Redux Component', () => {
  let store;
  let component;
  let data;

  beforeEach(() => {
    mockPlay.mockClear();

    store = mockStore(
      Map({
        itunes: fromJS({
          activeSong: undefined,
        }),
      }),
    );

    store.dispatch = jest.fn();

    data = Map({
      activeSong: fromJS({
        wrapperType: 'track',
        kind: 'song',
        artistId: 466532,
        collectionId: 1165630592,
        trackId: 1165631048,
        artistName: 'Korn',
        collectionName: 'Follow the Leader',
        trackName: 'Justin',
        collectionCensoredName: 'Follow the Leader',
        trackCensoredName: 'Justin',
        artistViewUrl: 'https://music.apple.com/us/artist/korn/466532?uo=4',
        collectionViewUrl:
          'https://music.apple.com/us/album/justin/1165630592?i=1165631048&uo=4',
        trackViewUrl:
          'https://music.apple.com/us/album/justin/1165630592?i=1165631048&uo=4',
        previewUrl:
          'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/54/57/70/545770b4-fa22-3d84-50fd-8007f97a4192/mzaf_8941792383747410636.plus.aac.p.m4a',
        artworkUrl30:
          'https://is1-ssl.mzstatic.com/image/thumb/Music124/v4/78/2b/8f/782b8fa1-0e66-b76d-83c9-80e6013e25f4/source/30x30bb.jpg',
        artworkUrl60:
          'https://is1-ssl.mzstatic.com/image/thumb/Music124/v4/78/2b/8f/782b8fa1-0e66-b76d-83c9-80e6013e25f4/source/60x60bb.jpg',
        artworkUrl100:
          'https://is1-ssl.mzstatic.com/image/thumb/Music124/v4/78/2b/8f/782b8fa1-0e66-b76d-83c9-80e6013e25f4/source/100x100bb.jpg',
        collectionPrice: 9.99,
        trackPrice: 1.29,
        releaseDate: '1998-08-18T12:00:00Z',
        collectionExplicitness: 'explicit',
        trackExplicitness: 'notExplicit',
        discCount: 1,
        discNumber: 1,
        trackCount: 14,
        trackNumber: 10,
        trackTimeMillis: 257547,
        country: 'USA',
        currency: 'USD',
        primaryGenreName: 'Hard Rock',
        isStreamable: true,
      }),
    });

    component = renderer.create(
      <Provider store={store}>
        <SongRowItem data={data} />
      </Provider>,
    );
  });

  it('should renders song info if data exist', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch setActiveSong on tap', () => {
    MusicPlayerHolder.setInstance(MusicControl);

    renderer.act(() => {
      component.toJSON().props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      ItunesActions.setActiveSong(data.toJS()),
    );
  });
});
