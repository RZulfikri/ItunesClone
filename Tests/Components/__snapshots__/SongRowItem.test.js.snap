// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongRowItem React-Redux Component should renders song info if data exist 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderBottomColor": "#E5E5E5",
      "borderBottomWidth": 1,
      "flexDirection": "row",
      "opacity": 1,
      "padding": 15,
    }
  }
>
  <Image
    resizeMethod="resize"
    resizeMode="cover"
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "height": 50,
        "width": 50,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "marginLeft": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }
    />
    <Text
      style={
        Object {
          "fontSize": 12,
        }
      }
    />
    <Text
      style={
        Object {
          "fontSize": 10,
        }
      }
    />
  </View>
  <View />
</View>
`;
